### PR TITLE
take fee when cancel order

### PR DESCRIPTION
**Pull request by nguyenbatam**
SELL 1 BTC => TOMO ,,
order.Quantity =1 & fee rate =2 => cancel fee =2/10000 BTC

BUY 1 BTC => TOMO with Price : 10000
quoteTokenQuantity = 10000 & fee rate =2 => cancel fee = 2 TOMO

RelayFee = 0.0001 TOMO

### CODE REVIEW

File: tomox/order_processor.go [hunk 1]
```diff
@@ -541,3 +543,67 @@ func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *tomox_stat
 	tomox_state.SetTokenBalance(makerExOwner, newMakerFee, makerOrder.QuoteToken, statedb)
 	return nil
 }
+
+func (tomox *TomoX) ProcessCancelOrder(tomoXstatedb *tomox_state.TomoXStateDB, statedb *state.StateDB, chain consensus.ChainContext, coinbase common.Address, orderBook common.Hash, order *tomox_state.OrderItem) (error, bool) {
+	exRelayerFee := tomox_state.GetExRelayerFee(order.ExchangeAddress, statedb)
```
Comment (thanhnguyennguyen) [inline tomox/order_processor.go]: The following function `GetExRelayerFee` doesn't return remaining fee of relayer.
It returns feeRate decimal. 
For example:
` _fee = 1` means `feeRate = 1/1000 = 0.1%`
To get remaining fee, let's get value at slot `_deposit`, then substract ` lockedFund = 20 000 TOMO`
```
func GetExRelayerFee(relayer common.Address, statedb *state.StateDB) *big.Int {
	slot := RelayerMappingSlot["RELAYER_LIST"]
	locBig := GetLocMappingAtKey(relayer.Hash(), slot)
	locBig = locBig.Add(locBig, RelayerStructMappingSlot["_fee"])
	locHash := common.BigToHash(locBig)
	return statedb.GetState(common.HexToAddress(common.RelayerRegistrationSMC), locHash).Big()
}
```

File: tomox/order_processor.go [hunk 2]
```diff
@@ -541,3 +543,67 @@ func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *tomox_stat
...
+	}
+	baseTokenDecimal, err := tomox.GetTokenDecimal(chain, statedb, coinbase, order.BaseToken)
+	if err != nil || baseTokenDecimal.Sign() == 0 {
+		log.Debug("Fail to get tokenDecimal ", "Token", order.BaseToken.String(), "err", err)
+		return err, false
+	}
+	var makerBalance *big.Int
+	switch order.Side {
+	case tomox_state.Ask:
+		makerBalance = tomox_state.GetTokenBalance(order.UserAddress, order.BaseToken, statedb)
```
Comment (thanhnguyennguyen) [inline tomox/order_processor.go]: :+1: 
in this case, user has baseToken, it makes sense to charge cancelFee in baseToken

File: tomox/order_processor.go [hunk 3]
```diff
@@ -541,3 +543,71 @@ func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *tomox_stat
 	tomox_state.SetTokenBalance(makerExOwner, newMakerFee, makerOrder.QuoteToken, statedb)
 	return nil
 }
+
+func (tomox *TomoX) ProcessCancelOrder(tomoXstatedb *tomox_state.TomoXStateDB, statedb *state.StateDB, chain consensus.ChainContext, coinbase common.Address, orderBook common.Hash, order *tomox_state.OrderItem) (error, bool) {
+	if err := tomox_state.CheckRelayerFee(order.ExchangeAddress, common.RelayerCancelFee, statedb); err != nil {
+		log.Debug("Relayer not enough fee when cancel order", "err", err)
+		return nil, true
```
Comment (thanhnguyennguyen) [inline tomox/order_processor.go:550]: return fmt.Errorf(....) ?

File: tomox/order_processor.go [hunk 4]
```diff
@@ -541,3 +543,71 @@ func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *tomox_stat
...
+	}
+	var tokenBalance *big.Int
+	switch order.Side {
+	case tomox_state.Ask:
+		tokenBalance = tomox_state.GetTokenBalance(order.UserAddress, order.BaseToken, statedb)
+	case tomox_state.Bid:
+		tokenBalance = tomox_state.GetTokenBalance(order.UserAddress, order.QuoteToken, statedb)
+	default:
+		log.Debug("Not found order side", "Side", order.Side)
+		return nil, true
```
Comment (thanhnguyennguyen) [inline tomox/order_processor.go:565]: return fmt.Errorf(....) ?

File: tomox/order_processor.go [hunk 5]
```diff
@@ -541,3 +543,71 @@ func DoSettleBalance(coinbase common.Address, takerOrder, makerOrder *tomox_stat
...
+	}
+	log.Debug("ProcessCancelOrder", "baseToken", order.BaseToken, "quoteToken", order.QuoteToken, "makerPrice", order.Price, "baseTokenDecimal", baseTokenDecimal, "quantity", order.Quantity)
+	feeRate := tomox_state.GetExRelayerFee(order.ExchangeAddress, statedb)
+	tokenCancelFee := getCancelFee(baseTokenDecimal, feeRate, order)
+	if tokenBalance.Cmp(tokenCancelFee) < 0 {
+		log.Debug("User not enough balance when cancel order", "Side", order.Side, "Price", order.Price, "Quantity", order.Quantity, "balance", tokenBalance, "fee", tokenCancelFee)
+		return nil, true
```
Comment (thanhnguyennguyen) [inline tomox/order_processor.go:572]: return fmt.Errorf(....) ?

**Review by thanhnguyennguyen — COMMENTED** [review]
Check deposit >= 20 000 TOMO

https://github.com/tomochain/tomochain/blob/master/tomox/tomox_state/relayer_state.go#L110
https://github.com/tomochain/tomochain/blob/master/tomox/tomox_state/relayer_state.go#L257

File: tomox/order_processor.go (hunk 5 above)
Comment (nguyenbatam) [reply tomox/order_processor.go:572]: reject order . Dont need error

File: tomox/order_processor.go (hunk 4 above)
Comment (nguyenbatam) [reply tomox/order_processor.go:565]: reject order . Dont need error

File: tomox/order_processor.go (hunk 3 above)
Comment (nguyenbatam) [reply tomox/order_processor.go:550]: reject order . Dont need error

**Comment by thanhnguyennguyen** [comment]
This will resolve #21

File: tomox/order_processor.go (hunk 3 above)
Comment (thanhnguyennguyen) [reply tomox/order_processor.go:550]: ok

**Review by thanhson1085 — APPROVED** [review]

**Pull request merged**